bundles/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/handler/AbstractOmnilinkStatusHandler.java

```diff
@@ -0,0 +1,100 @@
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.omnilink.internal.handler;
+
+import java.util.Optional;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.digitaldan.jomnilinkII.MessageTypes.statuses.Status;
+
+/**
+ * The {@link AbstractOmnilinkStatusHandler} defines some methods that can be used across
+ * the many different units exposed by the OmniLink protocol to retrive updated status information.
+ *
+ * @author Craig Hamilton - Initial contribution
+ */
+@NonNullByDefault
+public abstract class AbstractOmnilinkStatusHandler<T extends Status> extends AbstractOmnilinkHandler {
+    private final Logger logger = LoggerFactory.getLogger(AbstractOmnilinkStatusHandler.class);
+
+    public AbstractOmnilinkStatusHandler(Thing thing) {
+        super(thing);
+    }
+
+    private volatile Optional<T> status = Optional.empty();
+
+    @Override
+    public void initialize() {
+        updateHandlerStatus();
+    }
+
+    /**
+     * Attempt to retrieve an updated status for this handler type.
+     *
+     * @return Optional with updated status if possible, empty optional otherwise.
+     */
+    protected abstract Optional<T> retrieveStatus();
+
+    /**
+     * Update channels associated with handler
+     *
+     * @param t Status object to update channels with
+     */
+    protected abstract void updateChannels(T t);
+
+    /**
+     * Process a status update for this handler. This will dispatch updateChannels where appropriate.
+     *
+     * @param t Status to process.
+     */
+    public void handleStatus(T t) {
+        this.status = Optional.of(t);
+        updateChannels(t);
+    }
+
+    @Override
+    public void channelLinked(ChannelUID channelUID) {
+        logger.debug("Channel linked: {} for zone {}", channelUID, getThingNumber());
+        if (status.isPresent()) {
+            updateChannels(status.get());
+        }
+    }
+
+    @Override
+    public void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
+        super.bridgeStatusChanged(bridgeStatusInfo);
+        updateHandlerStatus();
+    }
+
+    private void updateHandlerStatus() {
+        Bridge bridge = getBridge();
+        if (bridge != null && bridge.getStatus() == ThingStatus.ONLINE) {
+            Optional<T> status = retrieveStatus();
+            if (status.isPresent()) {
+                handleStatus(status.get());
+                updateStatus(ThingStatus.ONLINE);
+            } else {
+                logger.debug("Received null staus update!");
+                updateStatus(ThingStatus.UNKNOWN);
```

Comment:
In case of an error the Thing should be set to offline. The message you are logging could be used as the `ThingStatusDetailMessage`. The framework will log it, then.